Adding multiple role support in navigation

sharidas opened the pull request:
The navigation manager should support multiple
roles. It was allowed to enter only one role
in the info.xml. With this change, the app
developer can provide multiple roles comma
separted for role.

Signed-off-by: Sujith H <sharidasan@owncloud.com>

<!--
Thanks for submitting a change to ownCloud!

This is the bug tracker for the Server component. Find other components at https://github.com/owncloud/core/blob/master/.github/CONTRIBUTING.md#guidelines

For fixing potential security issues please see https://owncloud.org/security/

To make it possible for us to get your change reviewed and merged please fill out below information carefully.

Please note that any kind of change first has to be submitted to the master branch which holds the next major version of ownCloud.

We will carefully discuss if your change can or has to be backported to stable branches.
-->

## Description
<!--- Describe your changes in detail -->
It would be nice to have multiple roles. One scenario that came to my attention was with users page. The users page should be accessible for both admin and group admin. Hence this PR helps for the same.

## Related Issue
<!--- This project only accepts pull requests related to open issues -->
<!--- If suggesting a new feature or change, please discuss it in an issue first -->
<!--- If fixing a bug, there should be an issue describing it with steps to reproduce -->
<!--- Please link to the issue here: -->
https://github.com/owncloud/core/issues/30922

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
It would be nice to have multiple roles. One scenario that came to my attention was with users page. The users page should be accessible for both admin and group admin. Hence this PR helps for the same.

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
- [x] Verified that single roles and multiple roles with users page.

## Screenshots (if appropriate):

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

PVince81 commented:
Interesting. cc @pmaier1 


@sharidas what are roles ? is it an alias for groups ?

I don't remember seeing any concept of role in info.xml, would be great if you could clarify this. Maybe provide steps to demonstrate the behavior before and after your PR.

sharidas commented:
When I tried using user_management app, I was trying to test if everything works as before. And to my surprise it was broken when user logged in as group admin. By broken I mean, the users page wont get displayed or cannot be accessed from the UI for group admin. Had a look at market app's info.xml which showed me https://github.com/owncloud/market/blob/master/appinfo/info.xml#L28. After debugging a bit figured out we could have single value for the role in info.xml. Significance of role is : to whom should the UI of the app should be displayed. So if the role is `admin` only user logged in as admin could see the UI. If the role is `sub-admin` then sub-admin user could see it.

This also brings a question what if an app's UI should be visible or usable for both admin and sub-admin. Hence extended the info.xml to use `role="admin,sub-admin"` ( ref: https://github.com/owncloud/user_management/pull/10/files#diff-224bb875c5d3cfcd7621135069013242R15). The change in the core would help app developers to provide comma separated role ( instead of single role ). Hope this make sense?

PVince81 commented:
Do you mean there is a regression where a group admin cannot use the users page any more ? If yes, please bisect to find when the regression was introduced and make a p1 ticket as this is quite important and could be a release blocker.

DeepDiver1975 commented:
> If yes, please bisect to find when the regression was introduced and make a p1 ticket as this is quite important and could be a release blocker.

no release blocker - this is master only

the sub.admin role was introduced by myself while moving the user management app out side of the core repo

the sub-admin role was never part of stable10 -> no release blocker

sharidas commented:
> no release blocker - this is master only

Indeed this is not a release blocker as this issue should be there in master only.

PVince81 commented:
we still need a "regression" label and priority somewhere. would be good to have a ticket and clarify that it is master only

sharidas committed:
Adding multiple role support in navigation

The navigation manager should support multiple
roles. It was allowed to enter only one role
in the info.xml. With this change, the app
developer can provide multiple roles comma
separted for role.

Signed-off-by: Sujith H <sharidasan@owncloud.com>

DeepDiver1975 commented on lib/private/NavigationManager.php:
```diff
@@ -145,13 +145,30 @@ private function init() {
+
+			$discontinue = true;
+			if ($this->isAdmin() && (in_array('admin', $role, true) === true)) {
+				$discontinue = false;
```
just continue in here

DeepDiver1975 commented on lib/private/NavigationManager.php:
```diff
@@ -145,13 +145,30 @@ private function init() {
+			/**
+			 * Can have multiple roles like role="admin,sub-admin"
+			 */
+			$role = explode(',', $role);
```
array elements need to be trimmed

DeepDiver1975 commented on tests/lib/NavigationManagerTest.php:
```diff
@@ -232,6 +236,22 @@ public function providesNavigationConfig() {
+				'icon' => '/apps/test/img/app.svg',
+				'name' => 'Test',
+				'active' => false
+			]], ['navigation' => ['@attributes' => ['role' => 'admin,sub-admin'], 'route' => 'test.page.index']], false, true]
```
please add test for 'admin, sub-admin' and 'sub-admin' and 'admin'

sharidas commented on lib/private/NavigationManager.php:
```diff
@@ -145,13 +145,30 @@ private function init() {
+
+			$shallContinue = false;
+			if ($this->isAdmin() && (in_array('admin', $role, true) === true)) {
+				$shallContinue = true;
```
Code readability improved. ( no inverted logic ).

sharidas commented on lib/private/NavigationManager.php:
```diff
@@ -145,13 +145,30 @@ private function init() {
+			 */
+			$role = explode(',', $role);
+			//Remove whitespace in the role elements
+			$role = array_map('trim', $role);
```
Trim the elements of array.

sharidas commented on tests/lib/NavigationManagerTest.php:
```diff
@@ -232,6 +236,54 @@ public function providesNavigationConfig() {
+				'icon' => '/apps/test/img/app.svg',
+				'name' => 'Test',
+				'active' => false
+			]], ['navigation' => ['@attributes' => ['role' => '   admin   '], 'route' => 'test.page.index']], true]
```
Added tests for "admin,sub-admin", "admin" and "sub-admin". Added tests for roles with spaces.
